internal/transport/http2_client_test.go

```diff
@@ -0,0 +1,172 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+ */
+package transport
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"net"
+	"testing"
+	"time"
+
+	"golang.org/x/net/http2"
+	"google.golang.org/grpc/resolver"
+)
+
+// ClientPreface is the HTTP/2 client preface string.
+var ClientPreface = []byte(http2.ClientPreface)
+
+// framerWriteSettings is serialized byte sequence of []http2.Setting{}
+var framerWriteSettings = []byte{0, 0, 0, 4, 0, 0, 0, 0, 0}
+
+// framerWindowUpdate is serialized byte sequence of http2.Setting{ID:  http2.SettingInitialWindowSize, Val: 14465}
+var framerWindowUpdate = []byte{0, 0, 4, 8, 0, 0, 0, 0, 0, 0, 0, 56, 129}
+
+type clientPrefaceConn struct {
+	net.Conn
+}
+
+type clientPrefaceLengthConn struct {
+	net.Conn
+}
+
+type framerWriteSettingsConn struct {
+	net.Conn
+}
+
+type framerWindowUpdateConn struct {
+	net.Conn
+}
+
+func (cp *clientPrefaceConn) Write(b []byte) (n int, err error) {
+	if bytes.Equal(b, ClientPreface) {
+		return 0, errors.New("force error for preface write")
+	}
+	return cp.Conn.Write(b)
+}
+
+func (cp *clientPrefaceConn) Close() error {
+	return cp.Conn.Close()
+}
+
+func dialerClientPrefaceWrite(_ context.Context, addr string) (net.Conn, error) {
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &clientPrefaceConn{Conn: conn}, nil
+}
+
+func (cpl *clientPrefaceLengthConn) Write(b []byte) (n int, err error) {
+	if bytes.Equal(b, ClientPreface) {
+		incorrectPreface := "INCORRECT PREFACE\r\n\r\n"
+		n, err = cpl.Conn.Write([]byte(incorrectPreface))
+		return n, err
+	}
+	return cpl.Conn.Write(b)
+
+}
+
+func dialerClientPrefaceLength(_ context.Context, addr string) (net.Conn, error) {
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &clientPrefaceLengthConn{Conn: conn}, nil
+}
+
+func (fws *framerWriteSettingsConn) Write(b []byte) (n int, err error) {
+	if bytes.Equal(b, framerWriteSettings) {
+		return 0, errors.New("force error for Framer write setting")
+	}
+	return fws.Conn.Write(b)
+}
+
+func dialerFramerWriteSettings(_ context.Context, addr string) (net.Conn, error) {
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &framerWriteSettingsConn{Conn: conn}, nil
+}
+
+func (fwu *framerWindowUpdateConn) Write(b []byte) (n int, err error) {
+	if bytes.Equal(b, framerWindowUpdate) {
+		return 0, errors.New("force error for windowupdate")
+	}
+	return fwu.Conn.Write(b)
+}
+
+func dialerFramerWriteWindowUpdate(_ context.Context, addr string) (net.Conn, error) {
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &framerWindowUpdateConn{Conn: conn}, nil
+}
+
+func (s) TestNewHTTP2ClientTarget(t *testing.T) {
+	tests := []struct {
+		name     string
+		opts     ConnectOptions
+		expected string
+	}{
+		{
+			name:     "client-preface-write",
```

Comment:
Seems like the error path is already covered by other tests. https://app.codecov.io/gh/grpc/grpc-go/blob/master/internal%2Ftransport%2Fhttp2_client.go#L425

Reply:
removed this test. as specified already covered.